### PR TITLE
Dynamic dashboard Enhancements

### LINKED ISSUE #973: Dynamic Dashboard
**Is your feature request related to a problem? Please describe.**
 The current organization dashboard lacks the ability to dynamically update and adapt to changing data and is not the best aesthetically

**Describe the solution you'd like**
Add a latest Post component to the dashboard
Add an upcoming Events component to the dashboard
Improving the look of the Statistics section of the dashboard

**Describe alternatives you've considered**
Including a notifications card to the component (Currently an unimplemented feature)

**Potential internship candidates**
Please read this if you are planning to apply for a Palisadoes Foundation internship https://github.com/PalisadoesFoundation/talawa/issues/359

### PR DESCRIPTION
**Pull request by JamarTG**
**What kind of change does this PR introduce?**

1. extended CardItem component to support the display of location and date for events and the creator for the posts as it is critical in the user deciding whether to view the full details of the previewed event/post. 
2. added location, date, and author icons to the CardItem Info for a more engaging look
3. two new functions have been introduced in the latest feed to handle the display of the five most recent posts and events.
4. added createdAt field to implement time display functionality for posts
5. dates have been changed to a more readable format

**Issue Number:**

Fixes #973

**Did you add tests for your changes?**

Yes

**Snapshots/Videos:**

image

https://github.com/PalisadoesFoundation/talawa-admin/assets/71823011/a83bff5e-ccaf-45d9-a295-7021732266f9

**Summary**

I had an implementation before but this change was made before mine so I'm slowly introducing some of my changes to the organization dashboard 

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-admin/blob/master/CONTRIBUTING.md)?**

Yes

### CODE REVIEW

File: src/utils/handleLatestFeed.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
+import type { InterfaceQueryOrganizationPostListItem } from './interfaces';
+import type { InterfaceQueryOrganizationEventListItem } from './interfaces';
+
+function sortandSliceEventsByStartDate(
```
Comment (beingnoble03) [inline src/utils/handleLatestFeed.ts:4]:
```suggestion
function sortAndSliceEventsByStartDate(
```

Comment (JamarTG) [reply src/utils/handleLatestFeed.ts]: > I guess there's a typo in the variable name here.

Yea, I made it more consistent as it was used in other places

**Review by jordan-barrett-jm — DISMISSED** [review]
Changes meet expectations. Feature fulfils requirements set out.

**Comment by beingnoble03** [comment]
@JamarTG please update the PR description. 

"What kind of change does this PR introduce?" This section doesn't clearly explain the changes you have made.

**Comment by palisadoes** [comment]
@JamarTG please add a video too, it will help greatly in the evaluation

**Comment by palisadoes** [comment]
@beingnoble03 The video is added

**Comment by JamarTG** [comment]
The video is added and the PR description is updated

**Comment by palisadoes** [comment]
@rishav-jha-mech This may have an impact on your design and current open PR

**Comment by rishav-jha-mech** [comment]
@palisadoes, Yes these changes would create a merge conflict for me, as for the Design part it can be improved, I have not covered adding icons and locations for events and posts in the Figma Design, I will modify the design so that it would look better.

**Comment by palisadoes** [comment]
@rishav-jha-mech I'd like to merge this PR as these are interesting features. @JamarTG has already had to refactor his code because of other PRs. How difficult would the recovery be for you?

**Comment by rishav-jha-mech** [comment]
@palisadoes I have seen the changed files, I can manage the Merge Conflict for this PR 🚀

**Review by rishav-jha-mech — APPROVED** [review]

File: src/screens/OrganizationDashboard/OrganizationDashboard.tsx [hunk 2]
```diff
@@ -190,16 +195,19 @@ function organizationDashboard(): JSX.Element {
...
-                        return (
-                          <CardItem
-                            type="Event"
-                            key={event._id}
-                            time={event.startDate}
-                            title={event.title}
-                          />
-                        );
-                      })
+                      sortAndSliceEventsByStartDate(upcomingEvents).map(
```
Comment (beingnoble03) [inline src/screens/OrganizationDashboard/OrganizationDashboard.tsx:198]: Why are events sorted here on the client-side?

You can use `eventsByOrganizationConnection` query to fetch posts from backend itself sorted by `startDate`.

File: src/screens/OrganizationDashboard/OrganizationDashboard.tsx [hunk 3]
```diff
@@ -229,12 +237,17 @@ function organizationDashboard(): JSX.Element {
                         <h6>No posts present</h6>
                       </div>
                     ) : (
-                      postData?.postsByOrganization.slice(0, 5).map((post) => {
+                      sortAndSlicePostsByCreatedAt(
```
Comment (beingnoble03) [inline src/screens/OrganizationDashboard/OrganizationDashboard.tsx:240]: Similarly, Why are posts sorted here on the client-side?

You can use `postsByOrganizationConnection` query to fetch posts from backend itself sorted by `createdAt` field.

**Review by beingnoble03 — CHANGES_REQUESTED** [review]
@JamarTG I think there is no need to create util `handleLatestField`. You can use `connection` queries available on `talawa-api` to get sorted posts and events.

Please use the `connection` queries instead of writing a new util. This simply increases the unnecessary code.

**Comment by palisadoes** [comment]
@JamarTG Please make these minor changes so that we can merge.

**Comment by JamarTG** [comment]
image

I've been done with the changes but I've been stuck on this test case for the past 2 days. Is there any reason why this mock shown above would not work? I've literally tried everything you could think of over this time and nothing seems to work. After hours of trying I think it needs another pair of eyes. 

@beingnoble03 
@rishav-jha-mech   
@palisadoes 
@jordan-barrett-jm

**Comment by beingnoble03** [comment]
@JamarTG Have you checked other tests where this particular query (`eventsByOrganizationConnection` here) was used? You can copy the mock data from that test.

**Comment by JamarTG** [comment]
> @JamarTG Have you checked other tests where this particular query (`eventsByOrganizationConnection` here) was used? You can copy the mock data from that test.

Yes, I have tried this. It still doesn't work. I only changed the title to Event 1 to match the test case and it's still failing

**Comment by beingnoble03** [comment]
@JamarTG check the variables in the mocked Query. The variable name should be exact same.

**Comment by JamarTG** [comment]
> @JamarTG check the variables in the mocked Query. The variable name should be exact same.


image
image


This is how it's used and defined

**Comment by JamarTG** [comment]
> @JamarTG check the variables in the mocked Query. The variable name should be exact same.

Tested my mock inside OrganizationEvents and it works. The implementations in OrganizationEvents and OrganizationDashboard are the same. Still doesn't work in my file

**Comment by beingnoble03** [comment]
@JamarTG please fix the failing test.

File: src/screens/OrganizationDashboard/OrganizationDashboardMocks.ts [hunk 4]
```diff
@@ -68,59 +67,112 @@ export const MOCKS = [
...
+        },
       },
     },
   },
   {
     request: {
-      query: ORGANIZATION_EVENT_LIST,
+      query: ORGANIZATION_EVENT_CONNECTION_LIST,
+      variables: {
+        organization_id: '123',
```
Comment (beingnoble03) [inline src/screens/OrganizationDashboard/OrganizationDashboardMocks.ts:160]: @JamarTG you have used `organization_id: '123'` here in the mocks. Have you assigned the organization_id as `123` in the tests?

**Review by jordan-barrett-jm — APPROVED** [review]
Approved

**Pull request merged**